---
title: "R assignment by Maliha"
output: 
  pdf_document: default
  html_document: default
date: "2024-03-10"
---

#I loaded the fang_et_al.txt and snp_positions.txt file where the first row will contain the variable names (by the arguement header=TRUE).

```{r}
library (tidyverse)
library(tidyr)
library(dplyr)
setwd("C:/Users/Maliha Jahan/Desktop/PhD/2nd Semester/546/R Assignment")
fang_et_al <- read.table("fang_et_al_genotypes.txt", header = TRUE, sep = "\t")
snp_positions <- read.table("snp_position.txt", header = TRUE, sep = "\t")
```

#Inspection of Fang_et_al_genotypes.txt

a. Data type is inspected using data frame and vector function.
b. Then head, tail, dimension and structure are inspected.
c. View funtion is used to see the data frame.

```{r}
head(fang_et_al) 
tail(fang_et_al) 
dim(fang_et_al) 
str(fang_et_al) 
names(fang_et_al) 
sapply(fang_et_al, class) 
View(fang_et_al)
file_info <- file.info("fang_et_al_genotypes.txt")
```

#Inspection of snp_position.txt

a. Data type is inspected using data frame and vector function.
b. Then head, tail, dimension and structure are inspected.
c. View funtion is used to see the data frame.

```{r}
head(snp_positions) 
tail(snp_positions) 
dim(snp_positions) 
str(snp_positions) 
names(snp_positions) 
View(snp_positions)
file_info_SNP <- file.info("snp_position.txt")
```
#Snp_postion.txt

1. Here I have created a new data frame by filtering SNP_ID, Chromosome and Position columns from snp_position.txt and named it new_snp_positions 
2. Checked the diemnsion(number of columns and rows) of new_snp_positions
3. Then I exported the data frame to a file naming new_snp_positions.

```{r}
new_snp_positions <- snp_positions[,c("SNP_ID", "Chromosome", "Position")] 
dim(new_snp_positions)
write.table(new_snp_positions, './new_snp_positions.txt')
```

#From fang_et_al Creating and Transposing maize files with maize genotypes:
1.To create maize genotype
2. Created a data frame by filtering third column "Group" using key words ZMMIL,ZMMLR and ZMMMR and named it maize
3. Inspected the dimesnsion of the new object maize.
4. Exported the data frame to a file naming maize.txt
5. Transposed maize and created a maize_transposed object
6.  Changed the Sample_ID to SNP_ID

```{r}
maize <- fang_et_al %>% filter(Group == "ZMMIL" | Group == "ZMMLR" | Group == "ZMMMR")
dim(maize)
write.table(maize, "./maize.txt")
maize_transposed <- t(maize[,-c(1:3)]) %>% as.data.frame()
maize_transposed$SNP_ID <- rownames(maize_transposed)
view(maize_transposed)
write.table(maize_transposed, "./maize_transposed.txt")
```

#From fang_et_al Creating and Transposing teosinte files:
1. Repeated the same steps as maize files to create and transpose teosinte genotype files

```{r}
teosinte <- fang_et_al %>% filter(Group == "ZMPBA" | Group == "ZMPIL" | Group == "ZMPJA")
dim(teosinte)
write.table(teosinte, "./teosinte.txt")
teosinte_transposed <- t(teosinte[,-c(1:3)]) %>% as.data.frame()
teosinte_transposed$SNP_ID <- rownames(teosinte_transposed)
view(teosinte_transposed)
write.table(teosinte_transposed, "./teosinte_transposed.txt")
```

#Joining Maize and Teosinte file with new_snp_positions file:

```{r}
maize_joined <- merge(new_snp_positions,maize_transposed,by = 'SNP_ID')
colnames(maize_joined)[4:ncol(maize_joined)] <- as.character(maize$Sample_ID)
maize_joined[,-c(1:3)] <- lapply(maize_joined[,-c(1:3)],as.character)
unique(maize_joined$Chromosome)
write.table(maize_joined, "./maize_joined.txt")
view(maize_joined)

teosinte_joined <- merge(new_snp_positions,teosinte_transposed,by = 'SNP_ID')
colnames(teosinte_joined)[4:ncol(teosinte_joined)] <- as.character(teosinte$Sample_ID)
teosinte_joined[,-c(1:3)] <- lapply(teosinte_joined[,-c(1:3)],as.character)
unique(teosinte_joined$Chromosome)
write.table(teosinte_joined, "./teosinte_joined.txt")
view(teosinte_joined)
```

#10 files (1 for each chromosome) with SNPs ordered based on increasing position values and with missing data encoded by this symbol: ?
#10 files (1 for each chromosome) with SNPs ordered based on decreasing position values and with missing data encoded by this symbol: -

```{r}
# Finding duplicate column names in the maize_joined dataframe
dup_cols <- names(maize_joined)[duplicated(names(maize_joined))]

# Printing out the duplicate column names to understand which are duplicated
print(dup_cols)

# Removing the duplicated columns by creating a unique set of column names

# Ensure unique column names in maize_joined
names(maize_joined) <- make.names(names(maize_joined), unique = TRUE)

# Checking if duplicates were resolved
anyDuplicated(names(maize_joined))

# Extract the desired chromosomes (assuming they are labeled 1 through 10)
desired_chromosomes <- as.character(1:10)

# Defining the process_chromosome function
process_chromosome <- function(df, chromosome, missing_value_increase, missing_value_decrease) {
  # Filter for the specific chromosome
  chr_data <- df %>% filter(Chromosome == chromosome)
  
  # Processing for increasing position values
  chr_data_inc <- chr_data %>%
    arrange(Position) %>%
    mutate(across(starts_with("SNP"), ~replace_na(as.character(.), missing_value_increase)))
  
  # Writing a file with increasing position values
  write_csv(chr_data_inc, sprintf("maize_chr%s_increasing.csv", chromosome))
  
  # Process for decreasing position values
  chr_data_dec <- chr_data %>%
    arrange(desc(Position)) %>%
    mutate(across(starts_with("SNP"), ~replace_na(as.character(.), missing_value_decrease)))
  
  # Write to a file with decreasing position values
  write_csv(chr_data_dec, sprintf("maize_chr%s_decreasing.csv", chromosome))
}

# Applying the function to each of the desired chromosomes
results <- lapply(desired_chromosomes, process_chromosome, df = maize_joined, 
                  missing_value_increase = "?", missing_value_decrease = "-")

# After running the lapply function, checking for one of the files again.
file.exists("maize_chr1_increasing.csv")
```

#10 files (1 for each chromosome) with SNPs ordered based on increasing position values and with missing data encoded by this symbol: ?
#10 files (1 for each chromosome) with SNPs ordered based on decreasing position values and with missing data encoded by this symbol: -

```{r}
# Finding duplicate column names in the maize_joined dataframe
dup_cols <- names(teosinte_joined)[duplicated(names(teosinte_joined))]

# Printing out the duplicate column names to understand which are duplicated
print(dup_cols)

# Removing the duplicated columns by creating a unique set of column names

# Ensure unique column names in maize_joined
names(teosinte_joined) <- make.names(names(teosinte_joined), unique = TRUE)

# Checking if duplicates were resolved
anyDuplicated(names(teosinte_joined))

# Extract the desired chromosomes (assuming they are labeled 1 through 10)
desired_chromosomes <- as.character(1:10)

# Defining the process_chromosome function
process_chromosome <- function(df, chromosome, missing_value_increase, missing_value_decrease) {
  # Filter for the specific chromosome
  chr_data <- df %>% filter(Chromosome == chromosome)
  
  # Processing for increasing position values
  chr_data_inc <- chr_data %>%
    arrange(Position) %>%
    mutate(across(starts_with("SNP"), ~replace_na(as.character(.), missing_value_increase)))
  
  # Writing a file with increasing position values
  write_csv(chr_data_inc, sprintf("teosinte_chr%s_increasing.csv", chromosome))
  
  # Process for decreasing position values
  chr_data_dec <- chr_data %>%
    arrange(desc(Position)) %>%
    mutate(across(starts_with("SNP"), ~replace_na(as.character(.), missing_value_decrease)))
  
  # Write to a file with decreasing position values
  write_csv(chr_data_dec, sprintf("teosinte_chr%s_decreasing.csv", chromosome))
}

# Applying the function to each of the desired chromosomes
results <- lapply(desired_chromosomes, process_chromosome, df = teosinte_joined, 
                  missing_value_increase = "?", missing_value_decrease = "-")

# After running the lapply function, checking for one of the files again.
file.exists("teosinte_chr1_increasing.csv")

```

#partII

```{r}
library(ggplot2)
# Summarize the count of non-missing SNPs per chromosome for maize

snp_positions %>% filter(Position < 10e1000) %>% 
  ggplot(aes(as.double(Chromosome))) + geom_bar(fill = 'blue', color = 'green') + geom_text(stat = 'count', aes(label = ..count..), vjust = -1) +
  scale_x_continuous(breaks = 1:10) + theme_replace() + ggtitle("SNPs count/CHR") + ylab('Number of SNPs') + xlab('CHR') 
ggsave('./Visualization/SNPs_number.png', create.dir = TRUE)


snp_positions %>% filter(Position < 10e1000) %>% 
  ggplot(aes(as.double(Position)/1000000)) + geom_histogram(aes(y = ..density..), color = 'green', fill = "green", alpha = 0.4, bins = 20) + geom_density(aes(as.double(Position)/1000000), color = "blue") + facet_wrap(~ as.double(Chromosome), scales = "free") + theme_replace() + ggtitle("Across Chromosomes distribution") +
  xlab('Position (Mb)') + ylab('SNP density')
ggsave(paste0("./Visualization/SNPs_distribution.png"))
```

##Comparing SNP distribution between maize and teosinte individuals

```{r}
maize_joined %>% 
  filter(Position < 10e1000) %>% ggplot(aes(as.double(Chromosome))) + geom_bar(fill = 'blue', color = 'green') + 
  geom_text(stat = 'count', aes(label = ..count..), vjust = -1) + scale_x_continuous(breaks = 1:10) + theme_replace() +
  ggtitle("SNPs count/CHR") + ylab('Number of SNPs') + xlab('CHR') 
ggsave('./Visualization/SNPs_number_maize.png')

teosinte_joined %>% filter(Position < 10e1000) %>%  ggplot(aes(as.double(Chromosome))) + geom_bar(fill = 'blue', color = 'green') + 
  geom_text(stat = 'count', aes(label = ..count..), vjust = -1) + scale_x_continuous(breaks = 1:10) + theme_replace() +
  ggtitle("SNPs count/CHR") + ylab('Number of SNPs') + xlab('CHR') 
ggsave('./Visualization/SNPs_number_teosinte.png')
```

##2.Missing data and amount of heterozygosity

###2a. Proportion of homozygous and Hets by sample and Plotting

```{r}
mutate_genotypes <- 
  fang_et_al %>% select(-JG_OTU, -Group) %>% pivot_longer(!Sample_ID) %>% 
  mutate(Allele = ifelse(value %in% c('C/C', 'G/G', 'A/A', 'T/T'), 'Homozygote', ifelse(value == '?/?', 'NA', 'Heterozygote')))  

mutate_genotypes %>% group_by(Sample_ID) %>%  count(Allele) %>% 
  ggplot(aes(fill = Allele, y = n, x = Sample_ID)) +
  geom_bar(position = "fill", stat = "identity") +
  ggtitle("Proportion of Homozygotes, Heterozygotes and Missing Data by Sample") +
  ylab('Proportion') +
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())
ggsave('./Visualization/Allelic_proportion_by_sample.png')
```

##2b. Proportion of homozygous and Hets by Group and Plotting

```{r}
mutate_groups <- 
  fang_et_al %>% select(-Sample_ID, -JG_OTU) %>%   
  pivot_longer(!Group) %>% mutate(Allele = ifelse(value %in% c('C/C', 'G/G', 'A/A', 'T/T'), 'Homozygote', ifelse(value == '?/?', 'NA', 'Heterozygote')))  

mutate_groups %>% group_by(Group) %>%  count(Allele) %>% 
  ggplot(aes(fill = Allele, y = n, x = Group)) +
  geom_bar(position = "fill", stat = "identity") +
  ggtitle("Proportion of Homozygotes, Heterozygotes and Missing Data by Group") +
  ylab('Proportion') 
ggsave('./Visualization/Allelic_Proportion_by_group.jpg')
```

##3 Own_Plot:Proportion of Heterozygous Allele
#Here I have plotted proportion of each Heterozygous Allele in different genotype group.

```{r}
mutate_groups %>% filter(Allele == "Heterozygote") %>% group_by(Group) %>%  count(value) %>% 
  ggplot(aes(fill = value, y = n, x = Group)) +
  geom_bar(position = "fill", stat = "identity") +
  ggtitle("Proportion Heterozygote sites in each group") +
  ylab('Proportion') 
ggsave('./Visualization/Heterozygous_distribution.png')
```


## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
